### PR TITLE
Implementing code metric statistics in prezento

### PR DESCRIPTION
**Pull request by alvarofernandoms**
Signed-off-by: Álvaro Fernando alvarofernandoms@gmail.com
Signed-off-by: Lucas Leite lucas.srl08@gmail.com
Signed-off-by: Maria Luciene marialucienefelix@gmail.com
Signed-off-by: Parley Martins parleypachecomartins@gmail.com

### CODE REVIEW

File: features/step_definitions/metric_configuration_steps.rb [hunk 1]
```diff
@@ -41,6 +41,7 @@
   expect(page).to have_content(@metric_configuration.metric.name)
   expect(page).to have_content(@metric_configuration.metric.code)
   expect(page).to have_content(@metric_configuration.weight)
+  expect(page).to have_content("Percentage of use")
```
Comment (rafamanzo) [inline features/step_definitions/metric_configuration_steps.rb]: Is this expectation necessary? It is not redundant with the one at  features/metric_configuration/create.feature on line 102?

File: config/locales/views/metric_configurations/en.yml [hunk 2]
```diff
@@ -36,6 +36,7 @@ en:
   reading_group_helper_html: "The %{href} associated with this metric."
   no_models: "There are no %{model} yet!"
   choose_metric: "Choose a metric from a Base Tool:"
+  percentage_of_use: "Percentage of use"
```
Comment (rafamanzo) [inline config/locales/views/metric_configurations/en.yml]: "Usage percentage" is a better translation.

**Comment by rafamanzo** [comment]
Please have a look at the unit and acceptance tests.

As well, we need to think about some kind of cache for the percentages. Otherwise KalibroConfigurations will get overloaded. Do you have any ideas about this?

**Comment by rafamanzo** [comment]
Now you can update kalibro_client to 0.2.0 (https://rubygems.org/gems/kalibro_client) so you can properly test your features.

**Comment by rafamanzo** [comment]
I would suggest a rebase from master. Part of the tests failling in your build should not fail with the latest code from there.

If you need some help, let us know :)

**Comment by alvarofernandoms** [comment]
Hi @rafamanzo,

Oui, oui! You're right, we'll do the rebase. But first we're working in the cache for the percentage value. You suggested that we should use the Fragment Caching in the view. So, I try this code below, but the result it's a little messy. The value does not update if I add another metric or if the time pass:

```ruby
<p>
<strong><%=t('usage_percentage') %>:</strong>
    <%= cache(:time_to_idle => 30.seconds) "#{@statistic} %" do%>
    <% end %>
</p>
```

Maybe is better to use the Rails.cache.fetch method on metric_configurations_controller. What do you think?

Regards!

**Comment by diegoamc** [comment]
Hi @alvarofernandoms.
I've been reading about [Fragment Caching](http://guides.rubyonrails.org/caching_with_rails.html#fragment-caching) and I think that maybe what we need is to create a helper method that decides if the cache should expire. Check the documentation and see if that gives you any insights.

If you have further doubts, please ask!

**Comment by rafamanzo** [comment]
Hi @alvarofernandoms ,

I would do something like this:

```erb
<%= cache("#{@metric_configuration.metric.metric_collector_name}#_{@metric_configuration.metric.name}_statistic") do %>
  <p>
    <strong><%=t('usage_percentage') %>:</strong> <%= "#{@statistic} %" %>
  </p>
<% end %>
```

Notice that now I have a unique key per metric which I can use to expire in the controller's `create` action like this:

```ruby
expire_fragment("#{@metric_configuration.metric.metric_collector_name}#_{@metric_configuration.metric.name}_statistic")
```

After `save` was successfully called.

What do you think?

**Comment by alvarofernandoms** [comment]
ping @rafamanzo, @diegoamc 

We use <tt>:expires_in</tt> to set the cache life. The code is (pretty much like yours):

```erb
<% cache("#{@metric_configuration.metric.metric_collector_name}_#{@metric_configuration.metric.name}_statistic", :expires_in => 20.minutes) do %>
  <p>
    <strong><%=t('usage_percentage') %>:</strong> <%= "#{@statistic} %" %>
  </p>
<% end %>
```

In our environment, the tests work quite well. The cache stored in <tt>tmp/cache/...</tt> only changes/expires if you add a new metric in given configuration.

**Comment by rafamanzo** [comment]
Hello @alvarofernandoms,

Sorry for the delay on this. Would you like to update your branch so we can go on with this?

**Comment by paulormm** [comment]
@vitorbaraujo , would you talk to @dylanguedes to the MES'2017 team proceed with this PR?

**Comment by DylanGuedes** [comment]
IMO, the MES team can take this one. It looks like they only need to write a few tests and check the best way to cache it, looking tangible for them.

**Comment by alvarofernandoms** [comment]
Hi @rafamanzo. It's ok, don't worry. I believe that the current MES team will handle it flawless.

Best regards!

**Pull request closed without merging**